File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

namespace protein_translation {
    
    std::string_view getCodonFromRNA(std::string_view rna, int index) {
        return std::string_view(&rna[3 * index], 3); 
    }
    
    std::vector<std::string> proteins(std::string rna) {
        std::map<std::string_view, std::string> codonToProtein = {
            {"AUG", "Methionine"},
            {"UUU", "Phenylalanine"}, {"UUC", "Phenylalanine"},
            {"UUA", "Leucine"}, {"UUG", "Leucine"},
            {"UCU", "Serine"}, {"UCC", "Serine"}, {"UCA", "Serine"}, {"UCG", "Serine"},
            {"UAU", "Tyrosine"}, {"UAC", "Tyrosine"},
            {"UGU", "Cysteine"}, {"UGC", "Cysteine"},
            {"UGG", "Tryptophan"},
            {"UAA", "STOP"}, {"UAG", "STOP"}, {"UGA", "STOP"}
        };

        std::vector<std::string> proteins;
        for (long unsigned int i = 0; i < rna.size() / 3; i++) {
            std::string_view sv = getCodonFromRNA(rna, i);
            std::string protein = codonToProtein[sv];
            
            if (protein == "STOP")
                break;

            proteins.push_back(protein);
        }
        
        return proteins;
    }

}  // namespace protein_translation

File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let translateCodon codon =
    match codon with
    | "AUG" -> "Methionine"
    | "UUU" | "UUC" -> "Phenylalanine"
    | "UUA" | "UUG" -> "Leucine"
    | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
    | "UAU" | "UAC" -> "Tyrosine"
    | "UGU" | "UGC" -> "Cysteine"
    | "UGG" -> "Tryptophan"
    | "UAA" | "UAG" | "UGA" -> "STOP"
    | _ -> "Unknown"


let proteins rna = 
    let rec mapProteins(sequence: string): string list= 
        let protein = translateCodon sequence.[..2]
        let n = sequence.Length
        match protein with
        | "STOP" -> []
        | _ -> 
            if n = 3 then [protein]
            elif n < 3 then []
            else [protein] @ mapProteins(sequence.[3..]) 
    
    rna |> mapProteins

